pkg/ccl/changefeedccl/changefeed_processors.go

```diff
@@ -1780,3 +1912,123 @@ func (f *schemaChangeFrontier) hasLaggingSpans(
 	}
 	return frontier.Add(lagThresholdNanos, 0).Less(f.latestTs)
 }
+
+// orderedRowMerger takes payloads of rows ordered ascending by updated
+// timestamp for different processors and flushes out elements to an underlying
+// sink in timestamp order by storing the sorted rows for each processor in
+// separate lists and keeping a min heap of the earliest entry in each row to
+// determine the earliest across all nodes .
+type orderedRowMerger struct {
+	// orderedRows stores a map of processorID -> sorted ascending list of pending
+	// rows to be emitted.
+	orderedRows map[int32][]jobspb.OrderedRows_Row
+
+	// minHeap is a min heap storing the head of every orderedRows list in order
+	// to determine the earliest row across all processors.  Entries in the
+	// minHeap still remain in the orderedRows list, as that allows us to just
+	// check the first element of an orderedRows entry for a processor to know if
+	// that processor already has its element in the min heap or not.
+	minHeap RowHeap
+
+	rowsBuffered       int
+	bytesBuffered      int
+	bufferedBytesLimit int
+
+	// Set only when using a core changefeed
+	passthroughBuf *encDatumRowBuffer
+	alloc          tree.DatumAlloc
+
+	frontier *schemaChangeFrontier
+	sink     EventSink
+	metrics  *sliMetrics
+}
+
+func (m *orderedRowMerger) Append(ctx context.Context, rows jobspb.OrderedRows) error {
+	backfillTs := m.frontier.BackfillTS()
+
+	for _, row := range rows.Rows {
+		m.bytesBuffered += row.Size()
+		m.rowsBuffered += 1
+
+		// If the row is part of an ongoing backfill it can be directly emitted
+		// without wasting cycles on the minHeap Push/Pop.
+		if row.Updated.Equal(backfillTs) {
+			err := m.emitRow(ctx, row)
+			if err != nil {
+				return err
+			}
+		} else {
+			if len(m.orderedRows[rows.SourceProcessorID]) == 0 {
+				heap.Push(&m.minHeap, RowHeapEntry{source: rows.SourceProcessorID, row: row})
+			}
+			m.orderedRows[rows.SourceProcessorID] = append(m.orderedRows[rows.SourceProcessorID], row)
+		}
+	}
+	if m.bytesBuffered > m.bufferedBytesLimit {
+		return errors.Newf("ordered row buffer exceeded maximum byte limit (%d > %d)", m.bytesBuffered, m.bufferedBytesLimit)
+	}
```

Comment:
This is the main area of concern, I don't know if the default value of 512MiB is enough if people can even have rows that are 1MB each.  We also might need a warning to set a lower resolved frequency / having the default value lower for total ordered feeds, since a high value would mean this gets tripped even in normal operation.